references/plugins/capcenter.go

```diff
@@ -285,3 +285,75 @@ func (g *GithubCenter) SyncCapabilityFromCenter() error {
 	fmt.Printf("successfully sync %d/%d from %s remote center\n", success, total, g.centerName)
 	return nil
 }
+
+// GetCapibilitiesFromCenter returns all capabilities from cap center
+func (g *GithubCenter) GetCapibilitiesFromCenter() ([]types.Capability, error) {
+	_, dirs, _, _ := g.client.Repositories.GetContents(g.ctx, g.cfg.Owner, g.cfg.Repo, g.cfg.Path, &github.RepositoryContentGetOptions{Ref: g.cfg.Ref})
+	dm, err := (&common.Args{}).GetDiscoveryMapper()
+	caps := make([]types.Capability, 0)
+	if err != nil {
+		return nil, err
+	}
+	for _, addon := range dirs {
+		if *addon.Type != "file" {
+			continue
+		}
+		fileContent, _, _, err := g.client.Repositories.GetContents(g.ctx, g.cfg.Owner, g.cfg.Repo, *addon.Path, &github.RepositoryContentGetOptions{Ref: g.cfg.Ref})
+		if err != nil {
+			return nil, err
+		}
+		var data = []byte(*fileContent.Content)
+		if *fileContent.Encoding == "base64" {
+			data, err = base64.StdEncoding.DecodeString(*fileContent.Content)
+			if err != nil {
+				return nil, fmt.Errorf("decode github content %s err %w", *fileContent.Path, err)
+			}
+		}
+		cap, err := ParseAndSyncCapability(dm, data)
+		if err != nil {
+			fmt.Printf("parse definition of %s err %v\n", *fileContent.Name, err)
+			continue
+		}
+		caps = append(caps, cap)
+	}
+	return caps, nil
+}
+
+// GetCapContentAndObject will return a certain yaml file's content and cap object
+func (g *GithubCenter) GetCapContentAndObject(addonName string) ([]byte, types.Capability, error) {
+	_, dirs, _, err := g.client.Repositories.GetContents(g.ctx, g.cfg.Owner, g.cfg.Repo, g.cfg.Path, &github.RepositoryContentGetOptions{Ref: g.cfg.Ref})
+	if err != nil {
+		return []byte{}, types.Capability{}, err
+	}
+	c := &common.Args{}
+	dm, err := c.GetDiscoveryMapper()
+	if err != nil {
+		return []byte{}, types.Capability{}, err
+	}
+	for _, addon := range dirs {
+		if *addon.Type != "file" {
+			continue
+		}
+		fileContent, _, _, err := g.client.Repositories.GetContents(g.ctx, g.cfg.Owner, g.cfg.Repo, *addon.Path, &github.RepositoryContentGetOptions{Ref: g.cfg.Ref})
+		if err != nil {
+			return []byte{}, types.Capability{}, err
+		}
+		var data = []byte(*fileContent.Content)
+		if *fileContent.Encoding == "base64" {
+			data, err = base64.StdEncoding.DecodeString(*fileContent.Content)
+			if err != nil {
+				return []byte{}, types.Capability{}, fmt.Errorf("decode github content %s err %w", *fileContent.Path, err)
+			}
+		}
+		cap, err := ParseAndSyncCapability(dm, data)
+		if err != nil {
+			fmt.Printf("parse definition of %s err %v\n", *fileContent.Name, err)
+			continue
+		}
```

Comment:
The logic, retrieving and processing each dirs from a github repo, is repeated by three functions now, which results in lots of duplicated codes. Is it possible to refactor it to make a walker func that can be re-used by multiple functions?

Reply:
That's good suggestion！I should do that.

Reply:
@captainroy-hy please review :)